src/Adapter/Product/Combination/Repository/CombinationRepository.php

```diff
@@ -335,20 +335,63 @@ private function getAttributesInformation(array $attributeIds, int $langId): arr
 
     /**
      * @param ProductId $productId
-     * @param array $filters
+     * @param array<string, mixed> $filters
      *
      * @return QueryBuilder
      */
     private function getCombinationsQueryBuilder(ProductId $productId, array $filters): QueryBuilder
     {
-        //@todo: filters are not handled.
         $qb = $this->connection->createQueryBuilder();
         $qb->from($this->dbPrefix . 'product_attribute', 'pa')
             ->where('pa.id_product = :productId')
-            ->orderBy('id_product_attribute', 'asc')
             ->setParameter('productId', $productId->getValue())
         ;
 
+        // filter by attributes
+        if (isset($filters['attribute_ids'])) {
+            $combinationIds = $this->getCombinationIdsByAttributeIds($productId, (array) $filters['attribute_ids']);
+            $qb->andWhere($qb->expr()->in('pa.id_product_attribute', ':combinationIds'))
+                ->setParameter('combinationIds', $combinationIds, Connection::PARAM_INT_ARRAY)
+            ;
+        }
+
+        $qb->orderBy('id_product_attribute', 'asc');
+
         return $qb;
     }
+
+    /**
+     * @param ProductId $productId
+     * @param int[] $attributeIds
+     *
+     * @return int[]
+     */
+    private function getCombinationIdsByAttributeIds(ProductId $productId, array $attributeIds): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb->select('pac.id_product_attribute')
+            ->from($this->dbPrefix . 'product_attribute_combination', 'pac')
+            ->leftJoin(
+                'pac',
+                $this->dbPrefix . 'product_attribute',
+                'pa',
+                'pac.id_product_attribute = pa.id_product_attribute'
+            )
+            ->where('pa.id_product = :productId')
+            ->setParameter('productId', $productId->getValue())
+            ->andWhere($qb->expr()->in('pac.id_attribute', ':attributeIds'))
+            ->setParameter('attributeIds', $attributeIds, Connection::PARAM_INT_ARRAY)
+            ->groupBy('pac.id_product_attribute')
+        ;
+
+        $results = $qb->execute()->fetchAll();
+
+        if (!$results) {
+            return [];
+        }
+
+        return array_map(function (array $result): int {
+            return (int) $result['id_product_attribute'];
+        }, $results);
+    }
```

Comment:
why do you think so?

Reply:
To be consistent with  https://github.com/PrestaShop/PrestaShop/tree/develop/src/Core/Grid/Query ?

Reply:
@zuk3975  What you did seems correct to me! I think it's just a question of how the code is organized.

Reply:
It's true we didn't think about taking profit of the `DoctrineQueryBuilderInterface` in our repositories 🤔

Although I think the Repository itself still needs the `Connection` and be able to create builders for specific functions that don't exist in legacy classes (or simply to avoid relying on them)

But I guess the pagination part could respect the interface we already have for regular Grids, this would allow someone to use the query builder independently and use it in a regular Grid maybe

So I would suggest to:
- create a `CombinationQueryBuilder` that respects the interface
- but still inject it in the Repository so that we can still rely on it being the source of all DB end points, even if internally it uses another service for pagination

I think it would be a nice refacto, but I'm not sure this PR is the right place to do it I guess we could refacto all the existing repositories we have with pagination this way

What do you guys think? @matks ?

Reply:
oh about the grid query builder indeed :thinking: Idk really, maybe we could use it instead, but we wouldn't be able to reuse the query anywhere out of grid context (thats all grid query builders issue though).
Idk, @jolelievre wdyt, should we move list related queries to CombinationGridQueryBuilder? It would also allow using query builder hooks for module devs?

Reply:
After some discussion we agree with you @PululuK We're gonna create a `CombinationQueryBuilder` and move this code in it instead of the Repository It'll be different from other grids though because it will be used in a CQRS command (at least for now)

We hesitated implementing a full Grid with its definition here but since it's not rendered like a regular "twig grid" it seems overkill, so we only implement the query part

Though we think this could be useful if we have an implementation of the Grids for Javascript, we've been talking about it for a long time now but we didn't have time to do it Maybe this will be the occasion to implement it for 1.7.9 and the combination listing might be a nice candidate for it